### ISSUE TITLE
How to replace and train a Custom Dataset Using the training and validation code?

### ISSUE THREAD

**Issue by APAE12138** [CASIA-IVA-Lab/FastSAM#248]
I’m currently learning how to work with training and validation code, but I’m unsure about how to replace the dataset with my own for training purposes. I have a labeled dataset that includes train, val, and test folders, but I'm not certain how to correctly configure the dataset paths and format.
     Does anyone have experience in replacing the dataset for training? I’d appreciate some guidance, such as how to correctly configure the data.yaml file and adjust the code to handle custom datasets. If you have any detailed examples or suggestions, that would be super helpful. Thank you all in advance!

    有大佬知道怎么跑training and validation 部分的代码吗？作者发的我有点看不懂QAQ。我想使用自己的数据集进行训练，但一直不知道怎么修改里面的代码。我看了一些相关的教程，根据coco数据集的形式标注好了 train、val 和 test ，但是跑代码的时候提示找不到标签位置，不能开始训练。有人知道怎么修改吗？非常感谢大家的帮助！

**Comment by JingLinWuu** [comment]
`我也遇到这个问题，我发现用yolo标注格式传入就好了，coco格式不行。但是训练出来的效果非常一般，不知道哪个环节有问题`
I encountered this problem as well. I found that it works fine when using the **YOLO annotation format** for input, but the **COCO format** does not. However, the training effect is very mediocre, and I have no idea which step went wrong.

**Comment by Armadillooutfitter** [comment]
✅ The MOST IMPORTANT thing (read this first)

👉 FastSAM itself does NOT contain training / validation code
👉 You train it using Ultralytics YOLOv8 (segmentation mode)
👉 FastSAM = YOLOv8-Seg model + prompt logic

So:

❌ You do NOT modify FastSAM’s Python code for training

✅ You ONLY prepare your dataset + data.yaml

✅ You train using the yolo command
Why your training fails (“cannot find labels”)

This error almost always means ONE of these is wrong:

Folder structure ❌

Label format ❌ (YOLOv8-seg ≠ COCO JSON)

data.yaml path ❌

Label file names don’t match image names ❌

✅ CORRECT DATASET STRUCTURE (YOLOv8-SEG)

Your dataset MUST look like this:

dataset/
├── images/
│   ├── train/
│   │   ├── 0001.jpg
│   │   ├── 0002.jpg
│   ├── val/
│   │   ├── 0101.jpg
│   ├── test/
│   │   ├── 0201.jpg
│
├── labels/
│   ├── train/
│   │   ├── 0001.txt
│   │   ├── 0002.txt
│   ├── val/
│   │   ├── 0101.txt
│   ├── test/
│   │   ├── 0201.txt


🚨 Images and labels must have the SAME filename

LABEL FORMAT (THIS IS CRITICAL)

FastSAM uses YOLOv8 Segmentation, NOT COCO JSON.

❌ COCO format (JSON) → WILL NOT WORK
✅ YOLOv8-seg (polygon txt) → REQUIRED

Example label file (0001.txt)
0 0.12 0.34 0.20 0.35 0.25 0.45 0.10 0.40

Format:

<class_id> x1 y1 x2 y2 x3 y3 ... xn yn


Coordinates are normalized [0–1]

One object per line

Polygon points, NOT bounding boxes

CORRECT data.yaml

Create a file like fastsam.yaml

path: /absolute/path/to/dataset
train: images/train
val: images/val
test: images/test

nc: 1
names: ['object']


⚠️ Use ABSOLUTE PATH (very important)

Example:

path: /home/yourname/data/my_dataset

HOW TO START TRAINING (THIS IS THE COMMAND)

You DO NOT run FastSAM’s Python files for training.

You run YOLOv8:

yolo task=segment \
  mode=train \
  model=FastSAM.pt \
  data=fastsam.yaml \
  imgsz=1024 \
  epochs=50 \
  batch=4
Validation

Validation runs automatically during training.

If you want standalone validation:

yolo task=segment \
  mode=val \
  model=runs/segment/train/weights/best.pt \
  data=fastsam.yaml

After training

Your trained model is here:

runs/segment/train/weights/best.pt


Use it in FastSAM inference:

python Inference.py \
  --model_path runs/segment/train/weights/best.pt \
  --img_path test.jpg

❌ What you DO NOT need to modify

❌ Inference.py

❌ prompt.py

❌ Any FastSAM source code

🧠 Why authors’ code is confusing

Because:

They reused Ultralytics YOLO

They didn’t explain it clearly in README

Training code is external, not in the repo
